File: android-app/app/src/main/java/com/zenulabidin/xposter/scheduler/services/GCSService.java
package com.zenulabidin.xposter.scheduler.services;

import android.content.Context;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.zenulabidin.xposter.scheduler.models.RepliesResponse;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class GCSService {
    private static final String TAG = "GCSService";
    private static final String BUCKET_NAME = "your-bucket-name";
    private static final String PROJECT_ID = "your-project-id";
    
    private final Context context;
    private final OkHttpClient httpClient;
    private final Gson gson;
    
    private String accessToken = null;
    private long tokenExpiry = 0;

    public GCSService(Context context) {
        this.context = context;
        this.gson = new Gson();
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(60, TimeUnit.SECONDS)
                .build();
    }

    public RepliesResponse fetchReplies(String currentETag) {
        try {
            String accessToken = getAccessToken();
            if (accessToken == null) {
                Log.e(TAG, "Failed to get access token");
                return null;
            }

            // Check for processing lock to avoid race conditions
            if (!checkAndClearStaleProcessingLock(accessToken)) {
                Log.d(TAG, "Another instance is processing, skipping fetch");
                return new RepliesResponse(new ArrayList<>(), null, currentETag, false);
            }

            // First check metadata to see if file changed
            String metadataUrl = String.format(
                "https://storage.googleapis.com/storage/v1/b/%s/o/replies.json?alt=json",
                BUCKET_NAME
            );

            Request metadataRequest = new Request.Builder()
                    .url(metadataUrl)
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .build();

            try (Response metadataResponse = httpClient.newCall(metadataRequest).execute()) {
                if (!metadataResponse.isSuccessful()) {
                    if (metadataResponse.code() == 404) {
                        Log.d(TAG, "Replies file not found in GCS");
                        return new RepliesResponse(new ArrayList<>(), null, null, false);
                    }
                    Log.e(TAG, "Metadata request failed: " + metadataResponse.code());
                    return null;
                }

                String metadataJson = metadataResponse.body().string();
                Type mapType = new TypeToken<java.util.Map<String, Object>>(){}.getType();
                java.util.Map<String, Object> metadata = gson.fromJson(metadataJson, mapType);
                
                String newETag = (String) metadata.get("etag");
                
                if (newETag != null && newETag.equals(currentETag)) {
                    Log.d(TAG, "ETag unchanged, no new data");
                    return new RepliesResponse(new ArrayList<>(), null, newETag, false);
                }

                // Download the actual file
                String downloadUrl = String.format(
                    "https://storage.googleapis.com/storage/v1/b/%s/o/replies.json?alt=media",
                    BUCKET_NAME
                );

                Request downloadRequest = new Request.Builder()
                        .url(downloadUrl)
                        .addHeader("Authorization", "Bearer " + accessToken)
                        .build();

                try (Response downloadResponse = httpClient.newCall(downloadRequest).execute()) {
                    if (!downloadResponse.isSuccessful()) {
                        Log.e(TAG, "Download request failed: " + downloadResponse.code());
                        return null;
                    }

                    String jsonContent = downloadResponse.body().string();
                    RepliesResponse response = gson.fromJson(jsonContent, RepliesResponse.class);
                    
                    if (response == null) {
                        response = new RepliesResponse();
                        response.replies = new ArrayList<>();
                    }
                    
                    response.etag = newETag;
                    response.hasChanges = true;
                    
                    Log.d(TAG, "Successfully fetched " + (response.replies != null ? response.replies.size() : 0) + " replies");
                    return response;
                }
            }
            
        } catch (Exception e) {
            Log.e(TAG, "Error fetching replies from GCS", e);
            return null;
        }
    }

    public List<String> fetchTopics() {
        try {
            String accessToken = getAccessToken();
            if (accessToken == null) {
                return getDefaultTopics();
            }

            String url = String.format(
                "https://storage.googleapis.com/storage/v1/b/%s/o/topics.json?alt=media",
                BUCKET_NAME
            );

            Request request = new Request.Builder()
                    .url(url)
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    if (response.code() == 404) {
                        Log.d(TAG, "Topics file not found, returning default topics");
                        return getDefaultTopics();
                    }
                    Log.e(TAG, "Topics request failed: " + response.code());
                    return getDefaultTopics();
                }

                String jsonContent = response.body().string();
                Type mapType = new TypeToken<java.util.Map<String, List<String>>>(){}.getType();
                java.util.Map<String, List<String>> topicsData = gson.fromJson(jsonContent, mapType);
                
                List<String> topics = topicsData.get("topics");
                return topics != null ? topics : getDefaultTopics();
            }
            
        } catch (Exception e) {
            Log.e(TAG, "Error fetching topics from GCS", e);
            return getDefaultTopics();
        }
    }

    public boolean uploadTopics(List<String> topics) {
        try {
            String accessToken = getAccessToken();
            if (accessToken == null) {
                return false;
            }

            java.util.Map<String, Object> topicsData = new java.util.HashMap<>();
            topicsData.put("topics", topics);
            topicsData.put("last_updated", new java.util.Date().toString());

            String jsonContent = gson.toJson(topicsData);

            String url = String.format(
                "https://storage.googleapis.com/upload/storage/v1/b/%s/o?uploadType=media&name=topics.json",
                BUCKET_NAME
            );

            RequestBody body = RequestBody.create(
                jsonContent,
                MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url(url)
                    .post(body)
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .addHeader("Content-Type", "application/json")
                    .build();

            try (Response response = httpClient.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Successfully uploaded topics to GCS");
                    return true;
                } else {
                    Log.e(TAG, "Topics upload failed: " + response.code());
                    return false;
                }
            }
            
        } catch (Exception e) {
            Log.e(TAG, "Error uploading topics to GCS", e);
            return false;
        }
    }

    private String getAccessToken() {
        if (accessToken != null && System.currentTimeMillis() < tokenExpiry) {
            return accessToken;
        }

        try {
            // Read service account key from assets
            InputStream keyStream = context.getAssets().open("key.json");
            
            // For simplicity, using a basic implementation
            // In production, use Google Auth libraries
            // This is a simplified version - you should use proper OAuth2 flow
            
            // For now, return null to indicate authentication needed
            Log.w(TAG, "Access token generation not implemented - please implement OAuth2 flow");
            return null;
            
        } catch (IOException e) {
            Log.e(TAG, "Error reading service account key", e);
            return null;
        }
    }

    private boolean checkAndClearStaleProcessingLock(String accessToken) {
        try {
            String metadataUrl = String.format(
                "https://storage.googleapis.com/storage/v1/b/%s/o/processing.lock?alt=json",
                BUCKET_NAME
            );

            Request metadataRequest = new Request.Builder()
                    .url(metadataUrl)
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .build();

            try (Response metadataResponse = httpClient.newCall(metadataRequest).execute()) {
                if (metadataResponse.code() == 404) {
                    return true; // No lock exists, proceed
                }
                
                if (!metadataResponse.isSuccessful()) {
                    Log.w(TAG, "Lock check failed: " + metadataResponse.code());
                    return true; // Assume no lock on error
                }

                String metadataJson = metadataResponse.body().string();
                Type mapType = new TypeToken<java.util.Map<String, Object>>(){}.getType();
                java.util.Map<String, Object> metadata = gson.fromJson(metadataJson, mapType);
                
                String timeCreated = (String) metadata.get("timeCreated");
                if (timeCreated != null) {
                    long lockAge = System.currentTimeMillis() - java.time.Instant.parse(timeCreated).toEpochMilli();
                    
                    // Clear stale locks older than 5 minutes
                    if (lockAge > 300000) {
                        deleteLock(accessToken);
                        Log.d(TAG, "Cleared stale processing lock");
                        return true;
                    } else {
                        return false; // Active lock exists
                    }
                }
                return true;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error checking processing lock", e);
            return true; // Proceed on error
        }
    }

    private void deleteLock(String accessToken) {
        try {
            String deleteUrl = String.format(
                "https://storage.googleapis.com/storage/v1/b/%s/o/processing.lock",
                BUCKET_NAME
            );

            Request deleteRequest = new Request.Builder()
                    .url(deleteUrl)
                    .delete()
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .build();

            try (Response response = httpClient.newCall(deleteRequest).execute()) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Successfully deleted processing lock");
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "Failed to delete processing lock", e);
        }
    }

    private List<String> getDefaultTopics() {
        List<String> topics = new ArrayList<>();
        topics.add("AI");
        topics.add("Crypto");
        topics.add("Bitcoin");
        topics.add("Ethereum");
        topics.add("Motivation");
        topics.add("Machine Learning");
        topics.add("Blockchain");
        topics.add("Self-Improvement");
        topics.add("Tech Innovation");
        topics.add("Programming");
        topics.add("Science");
        topics.add("Gaming");
        topics.add("Environment");
        topics.add("Finance");
        topics.add("Health & Wellness");
        return topics;
    }
}

File: android-app/app/src/main/java/com/zenulabidin/xposter/scheduler/services/GCSService.kt
package com.zenulabidin.xposter.scheduler.services

import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.zenulabidin.xposter.scheduler.models.RepliesResponse
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.util.Date
import java.util.concurrent.TimeUnit

class GCSService(private val context: Context) {

    companion object {
        private const val TAG = "GCSService"
        private const val BUCKET_NAME = "your-bucket-name"
        private const val PROJECT_ID = "your-project-id"
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()
    private var accessToken: String? = null
    private var tokenExpiry: Long = 0

    fun fetchReplies(currentETag: String?): RepliesResponse? {
        return try {
            val token = getAccessToken() ?: run {
                Log.e(TAG, "Failed to get access token")
                return null
            }

            // Check for processing lock to avoid race conditions
            if (!checkAndClearStaleProcessingLock(token)) {
                Log.d(TAG, "Another instance is processing, skipping fetch")
                return RepliesResponse(
                    replies = emptyList(),
                    etag = currentETag,
                    hasChanges = false
                )
            }

            // First check metadata to see if file changed
            val metadataUrl = "https://storage.googleapis.com/storage/v1/b/$BUCKET_NAME/o/replies.json?alt=json"

            val metadataRequest = Request.Builder()
                .url(metadataUrl)
                .addHeader("Authorization", "Bearer $token")
                .build()

            httpClient.newCall(metadataRequest).execute().use { metadataResponse ->
                if (!metadataResponse.isSuccessful) {
                    if (metadataResponse.code == 404) {
                        Log.d(TAG, "Replies file not found in GCS")
                        return RepliesResponse(
                            replies = emptyList(),
                            etag = null,
                            hasChanges = false
                        )
                    }
                    Log.e(TAG, "Metadata request failed: ${metadataResponse.code}")
                    return null
                }

                val metadataJson = metadataResponse.body?.string().orEmpty()
                val mapType = object : TypeToken<Map<String, Any>>() {}.type
                val metadata: Map<String, Any> = gson.fromJson(metadataJson, mapType)

                val newETag = metadata["etag"] as? String

                if (newETag != null && newETag == currentETag) {
                    Log.d(TAG, "ETag unchanged, no new data")
                    return RepliesResponse(
                        replies = emptyList(),
                        etag = newETag,
                        hasChanges = false
                    )
                }

                // Download the actual file
                val downloadUrl = "https://storage.googleapis.com/storage/v1/b/$BUCKET_NAME/o/replies.json?alt=media"

                val downloadRequest = Request.Builder()
                    .url(downloadUrl)
                    .addHeader("Authorization", "Bearer $token")
                    .build()

                httpClient.newCall(downloadRequest).execute().use { downloadResponse ->
                    if (!downloadResponse.isSuccessful) {
                        Log.e(TAG, "Download request failed: ${downloadResponse.code}")
                        return null
                    }

                    val jsonContent = downloadResponse.body?.string().orEmpty()
                    val response = gson.fromJson(jsonContent, RepliesResponse::class.java)
                        ?: RepliesResponse(replies = emptyList())

                    response.etag = newETag
                    response.hasChanges = true

                    Log.d(TAG, "Successfully fetched ${response.replies?.size ?: 0} replies")
                    response
                }
            }

        } catch (e: Exception) {
            Log.e(TAG, "Error fetching replies from GCS", e)
            null
        }
    }

    fun fetchTopics(): List<String> {
        return try {
            val token = getAccessToken() ?: return getDefaultTopics()

            val url = "https://storage.googleapis.com/storage/v1/b/$BUCKET_NAME/o/topics.json?alt=media"

            val request = Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer $token")
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    if (response.code == 404) {
                        Log.d(TAG, "Topics file not found, returning default topics")
                        return getDefaultTopics()
                    }
                    Log.e(TAG, "Topics request failed: ${response.code}")
                    return getDefaultTopics()
                }

                val jsonContent = response.body?.string().orEmpty()
                val mapType = object : TypeToken<Map<String, List<String>>>() {}.type
                val topicsData: Map<String, List<String>> = gson.fromJson(jsonContent, mapType)

                topicsData["topics"] ?: getDefaultTopics()
            }

        } catch (e: Exception) {
            Log.e(TAG, "Error fetching topics from GCS", e)
            getDefaultTopics()
        }
    }

    fun uploadTopics(topics: List<String>): Boolean {
        return try {
            val token = getAccessToken() ?: return false

            val topicsData = mapOf(
                "topics" to topics,
                "last_updated" to Date().toString()
            )

            val jsonContent = gson.toJson(topicsData)
            val url = "https://storage.googleapis.com/upload/storage/v1/b/$BUCKET_NAME/o?uploadType=media&name=topics.json"

            val body = jsonContent.toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url(url)
                .post(body)
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (response.isSuccessful) {
                    Log.d(TAG, "Successfully uploaded topics to GCS")
                    true
                } else {
                    Log.e(TAG, "Topics upload failed: ${response.code}")
                    false
                }
            }

        } catch (e: Exception) {
            Log.e(TAG, "Error uploading topics to GCS", e)
            false
        }
    }

    private fun getAccessToken(): String? {
        if (accessToken != null && System.currentTimeMillis() < tokenExpiry) {
            return accessToken
        }

        return try {
            // Read service account key from assets
            context.assets.open("key.json")

            // For simplicity, using a basic implementation
            // In production, use Google Auth libraries
            // This is a simplified version - you should use proper OAuth2 flow

            // For now, return null to indicate authentication needed
            Log.w(TAG, "Access token generation not implemented - please implement OAuth2 flow")
            null

        } catch (e: IOException) {
            Log.e(TAG, "Error reading service account key", e)
            null
        }
    }

    private fun checkAndClearStaleProcessingLock(accessToken: String): Boolean {
        return try {
            val metadataUrl = "https://storage.googleapis.com/storage/v1/b/$BUCKET_NAME/o/processing.lock?alt=json"

            val metadataRequest = Request.Builder()
                .url(metadataUrl)
                .addHeader("Authorization", "Bearer $accessToken")
                .build()

            httpClient.newCall(metadataRequest).execute().use { metadataResponse ->
                when (metadataResponse.code) {
                    404 -> true // No lock exists, proceed
                    !in 200..299 -> {
                        Log.w(TAG, "Lock check failed: ${metadataResponse.code}")
                        true // Assume no lock on error
                    }
                    else -> {
                        val metadataJson = metadataResponse.body?.string().orEmpty()
                        val mapType = object : TypeToken<Map<String, Any>>() {}.type
                        val metadata: Map<String, Any> = gson.fromJson(metadataJson, mapType)

                        val timeCreated = metadata["timeCreated"] as? String
                        if (timeCreated != null) {
                            val lockAge = System.currentTimeMillis() - java.time.Instant.parse(timeCreated).toEpochMilli()

                            // Clear stale locks older than 5 minutes
                            if (lockAge > 300000) {
                                deleteLock(accessToken)
                                Log.d(TAG, "Cleared stale processing lock")
                                true
                            } else {
                                false // Active lock exists
                            }
                        } else {
                            true
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error checking processing lock", e)
            true // Proceed on error
        }
    }

    private fun deleteLock(accessToken: String) {
        try {
            val deleteUrl = "https://storage.googleapis.com/storage/v1/b/$BUCKET_NAME/o/processing.lock"

            val deleteRequest = Request.Builder()
                .url(deleteUrl)
                .delete()
                .addHeader("Authorization", "Bearer $accessToken")
                .build()

            httpClient.newCall(deleteRequest).execute().use { response ->
                if (response.isSuccessful) {
                    Log.d(TAG, "Successfully deleted processing lock")
                }
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to delete processing lock", e)
        }
    }

    private fun getDefaultTopics(): List<String> = listOf(
        "AI",
        "Crypto",
        "Bitcoin",
        "Ethereum",
        "Motivation",
        "Machine Learning",
        "Blockchain",
        "Self-Improvement",
        "Tech Innovation",
        "Programming",
        "Science",
        "Gaming",
        "Environment",
        "Finance",
        "Health & Wellness"
    )
}